src/Common/Auth/OneTimeAuth.php

```diff
@@ -0,0 +1,388 @@
+        $access_ip = $ip ?: $_SERVER['REMOTE_ADDR'] ?? null;
+        $sql = "UPDATE `onetime_auth` SET `remote_ip` = ?, `last_accessed` = current_timestamp(), `access_count` = `access_count`+1 WHERE `pid` = ? AND `onetime_token` = ?";
+
+        return sqlQuery($sql, array($access_ip, $pid, $token));
+    }
+
+    /**
+     * @param $token
+     * @param $pid
+     * @return bool|array|null
+     */
+    public function getOnetime($token, $pid = null): bool|array|null
+    {
+        $sql = "SELECT * FROM `onetime_auth` WHERE `onetime_token` Like BINARY ? LIMIT 1";
+        $bind = [$token];
+        if ($pid) {
+            $bind = [$pid, $token];
+            $sql = "SELECT * FROM `onetime_auth` WHERE `pid` = ? AND `onetime_token` = ? LIMIT 1";
+        }
+
+        return sqlQuery($sql, $bind);
+    }
+
+    /**
+     * @param $token
+     * @param $redirect_token
+     * @return array
+     */
+    public function processOnetime($token, $redirect_token): array
+    {
+        $auth = $this->decodePortalOneTime($token, $redirect_token);
+        if (!empty($auth['error'] ?? null)) {
+            $this->systemLogger->error("Failed " . $auth['error']);
+            unset($auth);
+            throw new OneTimeAuthException(xlt("Authentication Failed! Contact administrator."));
+        }
+        $patientService = new PatientService();
+        $patient = $patientService->findByPid($auth['pid']);
+
+        // preserve session for target use
+        $_SESSION['pid'] = $auth['pid'];
+        $_SESSION['auth_pin'] = $auth['pin'];
+        $_SESSION['auth_scope'] = $this->scope;
+        $_SESSION['redirect_target'] = $auth['redirect'];
+        $_SESSION['onetime'] = $auth['portal_pwd'];
+        $_SESSION['patient_portal_onsite_two'] = 1;
+
+        // set up the other variables needed for the session interaction
+        // this was taken from portal/get_patient_info.php
+        $userService = new UserService();
+        $tmp = $userService->getUser($patient['providerID']);
+        $_SESSION['providerName'] = ($tmp['fname'] ?? '') . ' ' . ($tmp['lname'] ?? '');
+        $_SESSION['providerUName'] = $tmp['username'] ?? null;
+        $_SESSION['sessionUser'] = '-patient-';
+        $_SESSION['providerId'] = $patient['providerID'] ? $patient['providerID'] : 'undefined';
+        $_SESSION['ptName'] = $patient['fname'] . ' ' . $patient['lname'];
+        // never set authUserID though authUser is used for ACL!
+        $_SESSION['authUser'] = 'portal-user';
+        $_SESSION['portal_username'] = $auth['username']; // required by portal/handle_note.php
+        $_SESSION['portal_login_username'] = $auth['login_username']; // required by portal/handle_note.php
+        // Set up the csrf private_key (for the patient portal)
+        //  Note this key always remains private and never leaves server session. It is used to create
+        //  the csrf tokens.
+        CsrfUtils::setupCsrfKey();
+        $auth['redirect'] .= "&me=" . session_id();
```

Comment:
what is an example value for `$auth['redirect']` here? (especially interested in the value you use here that is breaking on windows)

Reply:
`
https://opensourcedemr.us/interface/modules/custom_modules/oe-module-comlink-telehealth/public/index-portal.php?action=launch_patient_session&pc_eid=35
`